### PR TITLE
[#12048] Migrate SubmitFeedbackResponseAction's Logic and Db methods

### PR DESCRIPTION
**Pull request by FergusMok**
Part of #12048.

This PR is part 1 of 2, and is related to #12720. Splitting it up as the PR is quite big.

This PR will handle the database and logic layer needed for the actual migration of SubmitFeedbackResponseAction.

### CODE REVIEW

File: src/it/java/teammates/it/storage/sqlapi/FeedbackResponseCommentsDbIT.java
Comment (jayasting98) [inline src/it/java/teammates/it/storage/sqlapi/FeedbackResponseCommentsDbIT.java:82]: If I am not wrong, I think the ID should not change. If it changes, it might be a bug. I realise now that there are also a few other places where it is assumed they can be updated; I think they should be removed.

File: src/it/java/teammates/it/storage/sqlapi/FeedbackResponseCommentsDbIT.java
Comment (jayasting98) [inline src/it/java/teammates/it/storage/sqlapi/FeedbackResponseCommentsDbIT.java:84]: There's a method in the parent class called `verifyEquals` which should handle this for us.

**Review by jayasting98 — CHANGES_REQUESTED** [review]
Thanks for the work, and thanks for keeping the PR relatively small :grin:. I think there are some things that were carried over from the Datastore-related code that should be different for the new SQL database. Otherwise, it should be fine.

File: src/main/java/teammates/storage/sqlentity/FeedbackResponse.java
Comment (NicolasCwy) [inline src/main/java/teammates/storage/sqlentity/FeedbackResponse.java:46]: New to this, why do you have to add `CascadeType.PERSIST` here?

File: src/main/java/teammates/storage/sqlapi/FeedbackResponseCommentsDb.java
Comment (NicolasCwy) [inline src/main/java/teammates/storage/sqlapi/FeedbackResponseCommentsDb.java:146]: Can I check why other attributes like `giver`, `isVisibilityFollowingFeedbackQuestion` and `isCommentFromFeedbackParticipant` don't have to be checked?

**Review by NicolasCwy — COMMENTED** [review]
Looks good but some clarifications

File: src/main/java/teammates/storage/sqlentity/FeedbackResponse.java
Comment (FergusMok) [reply src/main/java/teammates/storage/sqlentity/FeedbackResponse.java:46]: `CascadeType.PERSIST` will make sure that persisting the FeedbackResponse entity will also persist the associated FeedbackResponseComments within the entity.

File: src/main/java/teammates/storage/sqlapi/EntitiesDb.java
Comment (FergusMok) [reply src/main/java/teammates/storage/sqlapi/EntitiesDb.java:4]: I copied this as part of the migration from the old `EntitiesDb`, seems like they used `com.google.common.base.Objects` there too

File: src/it/java/teammates/it/storage/sqlapi/FeedbackResponseCommentsDbIT.java
Comment (FergusMok) [reply src/it/java/teammates/it/storage/sqlapi/FeedbackResponseCommentsDbIT.java:84]: Thanks, that is very useful

File: src/it/java/teammates/it/storage/sqlapi/FeedbackResponseCommentsDbIT.java
Comment (FergusMok) [reply src/it/java/teammates/it/storage/sqlapi/FeedbackResponseCommentsDbIT.java:82]: Yup, I missed this out, thanks

File: src/main/java/teammates/storage/sqlapi/FeedbackResponseCommentsDb.java
Comment (FergusMok) [reply src/main/java/teammates/storage/sqlapi/FeedbackResponseCommentsDb.java:146]: Good point, I copied them over and missed these attributes, will add them in

File: src/main/java/teammates/storage/sqlentity/FeedbackResponse.java
Comment (FergusMok) [reply src/main/java/teammates/storage/sqlentity/FeedbackResponse.java:46]: On second thoughts, I think automatically cascading the persist is not very performant for now. 

There are many occasions where the response is updated for other details. Manually cascading will also take advantage of the current `OPTIMIZED_SAVING_POLICY_APPLIED` we have.

File: src/main/java/teammates/storage/sqlapi/FeedbackResponseCommentsDb.java
Comment (FergusMok) [reply src/main/java/teammates/storage/sqlapi/FeedbackResponseCommentsDb.java]: Shifted to #12475

File: src/main/java/teammates/storage/sqlapi/EntitiesDb.java
Comment (FergusMok) [reply src/main/java/teammates/storage/sqlapi/EntitiesDb.java]: Shifted to https://github.com/TEAMMATES/teammates/pull/12745

File: src/it/java/teammates/it/storage/sqlapi/FeedbackResponseCommentsDbIT.java
Comment (FergusMok) [reply src/it/java/teammates/it/storage/sqlapi/FeedbackResponseCommentsDbIT.java]: Shifted to https://github.com/TEAMMATES/teammates/pull/12745

File: src/it/java/teammates/it/storage/sqlapi/FeedbackResponseCommentsDbIT.java
Comment (FergusMok) [reply src/it/java/teammates/it/storage/sqlapi/FeedbackResponseCommentsDbIT.java]: Shifted to https://github.com/TEAMMATES/teammates/pull/12745

File: src/main/java/teammates/sqllogic/core/FeedbackResponsesLogic.java
Comment (NicolasCwy) [inline src/main/java/teammates/sqllogic/core/FeedbackResponsesLogic.java:197]: I feel like this line causes the previous test to fail. Perhaps you can compare feedbackResponse with oldResponse instead before calling update? I think updateFeedbackResponse causes the merge to save the comments that have changed in your test

I tried referring to this but didn't 100% understand it
https://docs.jboss.org/hibernate/orm/6.1/userguide/html_single/Hibernate_User_Guide.html#pc-merge-gotchas

**Review by NicolasCwy — COMMENTED** [review]
Left some comments about the failing test

**Review by cedricongjh — APPROVED** [review]
other than @NicolasCwy's comments, LGTM

File: src/main/java/teammates/sqllogic/core/FeedbackResponsesLogic.java
Comment (FergusMok) [reply src/main/java/teammates/sqllogic/core/FeedbackResponsesLogic.java:197]: In the failing test, the function only calls this line, but does not call `               frcLogic.updateFeedbackResponseComment(oldResponseComment);`. This is unintended behavior, as we did not specify the comments to cascade (https://docs.jboss.org/hibernate/orm/6.1/userguide/html_single/Hibernate_User_Guide.html#pc-cascade). 

Right now, I'm still trying to find out why the cascading automatically happens even when we did not state for the cascade to happen. Furthermore, it's an issue as I don't think we're planning for automatic cascade to be intended behavior in the future for performance reasons.

File: src/it/java/teammates/it/sqllogic/core/FeedbackResponsesLogicIT.java
Comment (FergusMok) [reply src/it/java/teammates/it/sqllogic/core/FeedbackResponsesLogicIT.java:124]: Good idea, I'll use `@Test(enabled = false)` as it'll be cleaner compared to comments. 

I'll be working on a fix for this in the future, once `SubmitFeedbackResponseAction` is migrated.

File: src/main/java/teammates/sqllogic/core/FeedbackResponsesLogic.java
Comment (NicolasCwy) [reply src/main/java/teammates/sqllogic/core/FeedbackResponsesLogic.java:197]: Ah I see, didn't realise you removed the `Persist` from the previous PR review. Should we leave put this on hold till we figure this out?

Comment (FergusMok) [reply src/main/java/teammates/sqllogic/core/FeedbackResponsesLogic.java:197]: I was hoping to investigate this and get it merged with the original OPTIMIZED PR too, but we're rushing to merge `SubmitFeedbackResponseAction`'s migration as it's a large part of other E2E tests. These 2 issues do not contribute much to the behavior of the program, but more on the performance side, so we'll leave that to later to improve on.

**Review by NicolasCwy — APPROVED** [review]
Due to this blocking other PRs like the E2E tests lets merge this is, although we'll have to keep working on why the update feedback response cascades to feedback response comments, even if the feedback response does not change

image

**Pull request merged**
